### ISSUE TITLE
Getting "UnboundLocalError: local variable 'lb_ix' referenced before assignment" when using CounterFactual explainer

### ISSUE THREAD

**Issue by sug4ndh** [SeldonIO/alibi#186]
Hey,

When I use CounterFactual Explainer with KNeighboursClassifier, Decision Tree and even with SVM, I am getting "UnboundLocalError: local variable 'lb_ix' referenced before assignment". On some datasets when I use SVM classifier then it works. But on other datasets, it gives the same error with SVM also!  What could be causing this error?

Also when using CounterFactualProto Explainer with KNeighboursClassifier (n_neighbors >= 10) or Decision tree/ random Forest does not return any counterfactual explanation even after setting eps to (0.5,0.5). How much can I increase the eps argument? Additionally, are there any other parameters that might help in getting a counterfactual example when they are tuned?

Thanks.

**Comment by bihamta** [comment]
Hey, for your first question as far as I understood by working with the code, the error happens when the constraints of counterfactuals found have not been satisfied, maybe you have set the tol lower or target_proba higher than possible. It depends on the dataset since they have different accuracy so sometimes it is harder for the algorithm to find a counterfactual with your desired confidence.

**Comment by arnaudvl** [comment]
Hi @sug4ndh . For your first question the comment from @bihamta is spot on.
Regarding your second question, parameters that could be tuned:
* `c_init`: higher values put more weight on the loss term encouraging the predicted class of the counterfactual to be different from the original class. This is the loss term where the numerical model gradients are used.
* `c_steps`: increasing this value allows for more iterations to find a good range for `c_init`
* `max_iterations`: increasing this value allows for more iterations to find a counterfactual
* `theta`: higher values put more weight on the prototype loss term encouraging the counterfactual to be pushed towards the prototype in a different class than the original class

If you set `verbose` to True, you can also check the values of the different loss terms and numerical gradients. This is a very useful way to check whether the weights on the different loss terms are a bit balanced. If you see for instance that for your value of `eps` the numerical gradients are still very low and the value for the "attack loss term" (which uses the numerical gradients) is low as well, then you should probably increase `eps`.

**Comment by sug4ndh** [comment]
Thanks @bihamta and @arnaudvl!

**Comment by arnaudvl** [comment]
Our pleasure!

**Comment by sug4ndh** [comment]
Hey @bihamta and @arnaudvl, I tried changing the values of tol and target_proba and I am noticing a weird behaviour!?! 

With tol set to 0.5 and target_proba set to 0.5, the Counterfactual explainer is returning the same sample as the counterfactual example and reports that counterfactual prediction is also of the same class. If I either lower tol than 0.5 or make target_proba higher than .5 then I get "local variable 'lb_ix' referenced before assignment" error again.

**Comment by jklaise** [comment]
@sug4ndh when setting both `target_proba=0.5` and `tol=0.5`, the condition for a successful counterfactual is that any instance for which the model predicts a different class with `0 <= p <= 1` is a counterfactual, so because the original prediction assigns some probability mass to all classes, the original instance qualifies as a counterfactual. Here's the code that tests the condition for being a counterfactual:

https://github.com/SeldonIO/alibi/blob/f3a064817375c9ce4de117a43477d7144113580a/alibi/explainers/counterfactual.py#L340-L341

With the second issue, I'm unable to reproduce this, but it looks like the error comes from the following lines:

https://github.com/SeldonIO/alibi/blob/f3a064817375c9ce4de117a43477d7144113580a/alibi/explainers/counterfactual.py#L512-L513

Basically, the default starting value of `lam_init=0.1` which is the constant in front of the prediction part of the counterfactual loss (equivalent to `c_init` for `CounterFactualProto`, we will make these consistent soon...) is too small, so the first pass of the algorithm cannot find an appropriate range of `lam` to find counterfactuals. Trying to increase `lam_init` should take care of this.

**Comment by sug4ndh** [comment]
@jklaise Thanks! I will try to increase lam_limit and see how it performs. If I get into any issues, I will report it here again.
